### PR TITLE
feat(core): Add resource type and properties of all CfnResource constructs to tree.json

### PR DESCRIPTION
**Pull request by nija-at**
Also, switch the children node from an array to an object with each child object keyed on its id.

TODO:
- [ ] Support for tokens

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by shivlaks** [comment]
> This seems upside down...
> 
> I don't think `CfnResource`s should be special nor that the tree should encode the concept of `type` and `properties`. I believe the design we discussed was that each tree node has a set of key-value attributes and constructs can "contribute" to this attribute bag by implementing some interface. Then `CfnResource` implements the interface and contributes the attributes `aws.cloudformation.type` and `aws.cloudformation.properties` (which will be key/value pairs).
> 
> Something like:
> 
> ```ts
> interface ITreeAttributes {
>   treeAttributes: { [key: string]: any };
> }
> ```
> 
> And then:
> 
> ```
> class CfnResource implements ITreeAttributes {
>   public get treeAttributes() {
>     return {
>       'aws:cloudformation:type': this.resourceType,
>       'aws:cloudformation:properties': this.cfnProperties
>     }
>   }
> }
> ```

This is my miss for not closing the RFC sooner. Yes we did decide that we would have an attribute bag and constructs will provide information into it. L1's would report back all the properties in `toCloudFormation`. L2s would optionally contribute properties that are configurable, properties of interest, etc.

Type was the thing we hadn't closed discussion on. Initially, I felt it's nice to have as a top-level attribute of a Node, but I agree that since it's not applicable to every construct, that it's better off living in the attribute bag.

I'll update the RFC to make the change to nest type into the attribute bag of a tree node, rather than a top-level attribute.

File: packages/@aws-cdk/core/test/private/test.tree-metadata.ts [hunk 1]
```diff
@@ -22,41 +18,91 @@ export = {
       tree: {
         id: 'App',
         path: '',
-        children: [
-          {
+        children: {
+          Tree: {
             id: 'Tree',
-            path: 'Tree'
+            path: 'Tree',
```
Comment (shivlaks) [inline packages/@aws-cdk/core/test/private/test.tree-metadata.ts]: Are these trailing commas stripped out somewhere? I'm not sure if that's compliant with the JSON specification.
I think it is if you're using something like [JSON5](https://json5.org/)

File: packages/@aws-cdk/core/test/private/test.tree-metadata.ts [hunk 2]
```diff
@@ -22,41 +18,91 @@ export = {
...
+
+    test.deepEqual(readJson(assembly.directory, treeArtifact!.file), {
+      version: 'tree-0.1',
+      tree: {
+        id: 'App',
+        path: '',
+        children: {
+          Tree: {
+            id: 'Tree',
+            path: 'Tree',
```
Comment (shivlaks) [inline packages/@aws-cdk/core/test/private/test.tree-metadata.ts]: ditto (re: trailing comma)

File: packages/@aws-cdk/core/test/private/test.tree-metadata.ts [hunk 3]
```diff
@@ -22,41 +18,91 @@ export = {
...
+              mycfnresource: {
+                id: 'mycfnresource',
+                path: 'mystack/mycfnresource',
+                type: { cfnResourceType: 'CDK::UnitTest::MyCfnResource' },
+                properties: {
+                  mystringpropkey: 'mystringpropval',
+                  mylistpropkey: [ 'listitem1' ],
+                  mystructpropkey: {
+                    myboolpropkey: true,
+                    mynumpropkey: 50,
```
Comment (shivlaks) [inline packages/@aws-cdk/core/test/private/test.tree-metadata.ts]: ditto

File: packages/@aws-cdk/core/test/private/test.tree-metadata.ts (hunk 1 above)
Comment (nija-at) [reply packages/@aws-cdk/core/test/private/test.tree-metadata.ts]: It is valid Javascript and ECMAScript - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas

Comment (shivlaks) [reply packages/@aws-cdk/core/test/private/test.tree-metadata.ts]: awesome, thanks for confirming! (was failing my linter, but looks like my linter validates compliance against pre ES5 as well

**Pull request closed without merging**